File: desktop-app/conflict_diff.test.js
const test = require("node:test");
const assert = require("node:assert/strict");
const D = require("./conflict_diff");

test("splitLines: leerer/null input → []", () => {
  assert.deepEqual(D.splitLines(""), []);
  assert.deepEqual(D.splitLines(null), []);
});

test("splitLines: mehrzeilig, \\r\\n und \\n", () => {
  assert.deepEqual(D.splitLines("a\nb\nc"), ["a", "b", "c"]);
  assert.deepEqual(D.splitLines("a\r\nb"), ["a", "b"]);
});

test("computeSideBySide: identische strings → alle 'equal'", () => {
  const rows = D.computeSideBySide("a\nb", "a\nb");
  assert.equal(rows.length, 2);
  assert.ok(rows.every((r) => r.kind === "equal"));
});

test("computeSideBySide: einzeilige änderung → 'change'", () => {
  const rows = D.computeSideBySide("A", "B");
  assert.deepEqual(rows, [{ kind: "change", left: "A", right: "B" }]);
});

test("computeSideBySide: zeile nur lokal → 'remove'", () => {
  const rows = D.computeSideBySide("a\nb", "a");
  assert.equal(rows.length, 2);
  assert.equal(rows[0].kind, "equal");
  assert.equal(rows[1].kind, "remove");
  assert.equal(rows[1].left, "b");
  assert.equal(rows[1].right, null);
});

test("computeSideBySide: zeile nur remote → 'add'", () => {
  const rows = D.computeSideBySide("a", "a\nb");
  assert.equal(rows[1].kind, "add");
  assert.equal(rows[1].right, "b");
  assert.equal(rows[1].left, null);
});

test("computeSideBySide: gemischt — change + remove", () => {
  const rows = D.computeSideBySide("erst\nalt\ndritt", "erst\nneu");
  assert.equal(rows[0].kind, "equal");
  assert.equal(rows[1].kind, "change");
  assert.equal(rows[1].left, "alt");
  assert.equal(rows[1].right, "neu");
  assert.equal(rows[2].kind, "remove");
  assert.equal(rows[2].left, "dritt");
});

test("hasChanges: equal-only → false", () => {
  assert.equal(D.hasChanges(D.computeSideBySide("x", "x")), false);
});

test("hasChanges: mit unterschied → true", () => {
  assert.equal(D.hasChanges(D.computeSideBySide("x", "y")), true);
});

test("computeSideBySide: beide leer → []", () => {
  assert.deepEqual(D.computeSideBySide("", ""), []);
});

test("computeSideBySide: nur lokal befüllt, remote leer → alles remove", () => {
  const rows = D.computeSideBySide("a\nb", "");
  assert.equal(rows.length, 2);
  assert.ok(rows.every((r) => r.kind === "remove"));
  assert.equal(rows[0].left, "a");
  assert.equal(rows[1].left, "b");
});

File: mobile-app/test/features/conflicts/conflict_diff_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:projectgamma_mobile/features/conflicts/conflict_diff.dart';

void main() {
  test('splitLines: leer/null → []', () {
    expect(splitLines(''), isEmpty);
    expect(splitLines(null), isEmpty);
  });

  test('splitLines: mehrzeilig, akzeptiert \\r\\n', () {
    expect(splitLines('a\nb\nc'), ['a', 'b', 'c']);
    expect(splitLines('a\r\nb'), ['a', 'b']);
  });

  test('computeSideBySide: identisch → alle equal', () {
    final rows = computeSideBySide('a\nb', 'a\nb');
    expect(rows.length, 2);
    expect(rows.every((r) => r.kind == DiffKind.equal), isTrue);
  });

  test('computeSideBySide: einzeilige änderung → change', () {
    final rows = computeSideBySide('A', 'B');
    expect(rows.length, 1);
    expect(rows.first.kind, DiffKind.change);
    expect(rows.first.left, 'A');
    expect(rows.first.right, 'B');
  });

  test('computeSideBySide: zeile nur lokal → remove', () {
    final rows = computeSideBySide('a\nb', 'a');
    expect(rows[1].kind, DiffKind.remove);
    expect(rows[1].left, 'b');
    expect(rows[1].right, isNull);
  });

  test('computeSideBySide: zeile nur remote → add', () {
    final rows = computeSideBySide('a', 'a\nb');
    expect(rows[1].kind, DiffKind.add);
    expect(rows[1].right, 'b');
    expect(rows[1].left, isNull);
  });

  test('computeSideBySide: gemischt change+remove', () {
    final rows = computeSideBySide('erst\nalt\ndritt', 'erst\nneu');
    expect(rows[0].kind, DiffKind.equal);
    expect(rows[1].kind, DiffKind.change);
    expect(rows[1].left, 'alt');
    expect(rows[1].right, 'neu');
    expect(rows[2].kind, DiffKind.remove);
    expect(rows[2].left, 'dritt');
  });

  test('hasChanges', () {
    expect(hasChanges(computeSideBySide('x', 'x')), isFalse);
    expect(hasChanges(computeSideBySide('x', 'y')), isTrue);
  });

  test('computeSideBySide: beide leer → []', () {
    expect(computeSideBySide('', ''), isEmpty);
  });

  test('computeSideBySide: remote leer → alles remove', () {
    final rows = computeSideBySide('a\nb', '');
    expect(rows.length, 2);
    expect(rows.every((r) => r.kind == DiffKind.remove), isTrue);
  });
}
